File: 2015/07.py
"""--- Day 7: Some Assembly Required ---"""

import re

def process_instructions(instructions):
  wires = {}

  for line in instructions:
    left, right = map(str.strip, line.split("->"))
    # lx -> x, 123 -> x
    m = re.findall("^([a-z]+|[0-9]+)$", left)
    if m:
      inputs = m[0]
      gate = "WIRE"
    # x OR y -> e, 1 OR z
    m = re.findall("(AND|OR)", left)
    if m:
      gate = m[0]
      inputs = re.findall("([a-z]+|[0-9]+)", left)
    # NOT x -> h
    m = re.findall("(NOT) ([a-z]+)$", left)
    if m:
      gate, inputs = m[0]
    # y RSHIFT 2 -> g
    m = re.findall("([a-z]+) (LSHIFT|RSHIFT) ([0-9]+)", left)
    if m:
      inputs = m[0][0], m[0][2]
      gate = m[0][1].strip()

    wires[right] = {"gate" : gate, "inputs" : inputs}
  return wires

def solve(instructions, find):

  wires = process_instructions(instructions)
  values = {}
  # bruteforce
  while find not in values.keys():
    for wire, circuit in wires.items():
      if wire not in values:
        if circuit["gate"] == "WIRE":
          if re.match("[0-9]+", circuit["inputs"]):
            values[wire] = int(circuit["inputs"])
          elif circuit["inputs"] in values:
            values[wire] = values[circuit["inputs"]]
        elif circuit["gate"] == "LSHIFT":
          if circuit["inputs"][0] in values:
            values[wire] = values[circuit["inputs"][0]] << int(circuit["inputs"][1])
        elif circuit["gate"] == "RSHIFT":
          if circuit["inputs"][0] in values:
            values[wire] = values[circuit["inputs"][0]] >> int(circuit["inputs"][1])
        elif circuit["gate"] == "NOT":
          if re.match("[0-9]+", circuit["inputs"]):
            values[wire] = ~ int(circuit["inputs"])
          elif circuit["inputs"] in values:
            values[wire] = ~ values[circuit["inputs"]]
        elif circuit["gate"] == "AND":
          if re.match("[0-9]+", circuit["inputs"][0]):
            # number & value
            if circuit["inputs"][1] in values:
              values[wire] = int(circuit["inputs"][0]) & values[circuit["inputs"][1]]
          elif circuit["inputs"][0] in values and circuit["inputs"][1] in values:
            values[wire] = values[circuit["inputs"][0]] & values[circuit["inputs"][1]]
        elif circuit["gate"] == "OR":
          if re.match("[0-9]+", circuit["inputs"][0]):
            # number & value
            if circuit["inputs"][1] in values:
              values[wire] = int(circuit["inputs"][0]) | values[circuit["inputs"][1]]
          elif circuit["inputs"][0] in values and circuit["inputs"][1] in values:
            values[wire] = values[circuit["inputs"][0]] | values[circuit["inputs"][1]]

  return values[find] % 65536

with open("input/07.txt") as f:
  print(solve(f, "a"))

with open("input/07_2.txt") as f:
  print(solve(f, "a"))

File: 2015/07.go
package main

import (
	"bufio"
	"errors"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

type inst struct {
	instruction string
	left        []string
	output      string
}

func format(line string) inst {
	split := strings.Split(line, " -> ")
	output := split[1]
	command := strings.Split(split[0], " ")
	// e.g. 1 -> a
	if len(command) == 1 {
		return inst{"ASSIGN", command[0:1], output}
	}
	// e.g. NOT b -> a
	if len(command) == 2 {
		return inst{command[0], command[1:2], output}
	}
	// e.g. x LSHIFT y -> z
	return inst{command[1], []string{command[0], command[2]}, output}
}

func lookup(circuit map[string]uint16, value string) (uint16, error) {

	if strings.ContainsAny(value, "0123456789") {
		res, err := strconv.ParseUint(value, 10, 16)
		if err != nil {
			log.Fatal(err)
		}
		return uint16(res), nil
	}

	if res, ok := circuit[value]; ok {
		return res, nil
	}
	return 0, errors.New("Not found in map")
}

func lshift(x uint16, y uint16) uint16 {
	return x << y
}

func rshift(x uint16, y uint16) uint16 {
	return x >> y
}

func and(x uint16, y uint16) uint16 {
	return x & y
}

func or(x uint16, y uint16) uint16 {
	return x | y
}

func not(x uint16) uint16 {
	return (0 - x) - 1
}

func process(circuit map[string]uint16, instructions []inst) map[string]uint16 {

	for _, instruction := range instructions {
		output := instruction.output
		if _, ok := circuit[output]; ok {
			// skip if we already have decoded the output
		} else {
			wire1, err := lookup(circuit, instruction.left[0])
			if err != nil {
				continue
			}
			var fn func(uint16, uint16) uint16
			switch instruction.instruction {
			case "ASSIGN":
				circuit[output] = wire1
				continue
			case "NOT":
				circuit[output] = not(wire1)
				continue
			case "LSHIFT":
				fn = lshift
			case "RSHIFT":
				fn = rshift
			case "AND":
				fn = and
			case "OR":
				fn = or
			default:
				fmt.Println("Unknown command", instruction.instruction)
				continue
			}
			wire2, err := lookup(circuit, instruction.left[1])
			if err != nil {
				continue
			}
			circuit[output] = fn(wire1, wire2)
		}
	}
	return circuit
}

func iterate(circuit map[string]uint16, instructions []inst) uint16 {
	for {
		if a, ok := circuit["a"]; ok {
			return a
		}
		circuit = process(circuit, instructions)
	}
}

func main() {
	file, err := os.Open("input/07.txt")
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)

	var instructions []inst

	for scanner.Scan() {
		res := format(scanner.Text())
		instructions = append(instructions, res)
	}
	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}

	a := iterate(make(map[string]uint16), instructions)
	fmt.Println(a)

	instructions = append([]inst{inst{"ASSIGN", []string{strconv.Itoa(int(a))}, "b"}},
		instructions...)

	fmt.Println(iterate(make(map[string]uint16), instructions))
}
